quickwit-metastore/src/lib.rs

```diff
@@ -59,3 +63,68 @@ pub use metastore::{
     IndexMetadata, MetadataSet, Metastore, SplitMetadata, SplitMetadataAndFooterOffsets, SplitState,
 };
 pub use metastore_resolver::{MetastoreFactory, MetastoreUriResolver};
+
+/// A trait to allow for our data model to check itself
+#[async_trait]
+pub trait Check {
+    /// An async function that implements the check on this type
+    async fn check(&self) -> Result<(), anyhow::Error>;
+}
+
+#[cfg(any(test, feature = "testsuite"))]
+#[async_trait]
+impl Check for MockMetastore {
+    async fn check(&self) -> anyhow::Result<()> {
+        Ok(())
+    }
+}
+
+/// Decorate a result as a check (for use in [`do_checks`]).
+///
+/// This takes out the original error of the `Result` (if any), replacing it with "taken".
+/// This measure is unfortunately required by the fact that `anyhow::Error` is not `Clone`.
+/// As the checks should bail anyway, this should not be a problem.
+pub fn checked<T>(r: &mut anyhow::Result<T>) -> Pin<Box<(dyn Future<Output = anyhow::Result<()>> + Send)>> {
+    let res = match r {
+        Ok(_) => Ok(()),
+        Err(e) => Err(std::mem::replace(e, anyhow::anyhow!("taken"))),
+    };
+    Box::pin(std::future::ready(res))
+}
+
+/// Decorate a check with nice output
+pub fn map_results<F: Future<Output = anyhow::Result<()>>>(
```

Comment:
This function takes a future, and prints stuff in the stderr when it gets resolved.

The name of the function does not tell anything about what it does.

Reply:
As mentionned in the issue, we will want to use this check operation in different contexts.
The output will be the CLI today. In the future, we will very likely want to display its output in an admin UI.

That's where I would have preferred to have a tiny bit of overengineering and have a function that returns the outcome of the checks... 
And another that displays the outcome of the checks.

Here this function decorates -in the sense of a python decorator- the check function future.
It has the benefit of giving a cool CLI experience where each check appears asynchronously one ofter the other on stderr, but it is not as nice from the code POV.

Reply:
I'm curious... Is there a reason, why this is not using async?

Reply:
Because the checks may wait on certain things like locks, database connections, etc., which is a good use of the async runtime we already have. We might even make it multithreaded, but I don't think this will win us much.